### PR TITLE
OnlyRewriteDeprecatedURLs and AlwaysRewriteDistributionURLs

### PR DESCRIPTION
**Pull request by sphen13**
Add in two new options:

`OnlyRewriteDeprecatedURLs`

> This boolean preference allows for you to choose to only rewrite deprecated product URLs so they point to your LocalCatalogURLBase. Products currently available in Apple's catalogs will continue to point to Apple's CDN. This preference is intended to allow you to save bandwidth.
Note: Setting this preference requires LocalCatalogURLBase to be set and will locally cache all packages even if they will not be referenced in your catalogs.

`AlwaysRewriteDistributionURLs`

> This boolean preference allows for rewriting of the DistributionURLs inside the catalogs so they point to the LocalCatalogURLBase. This preference is intended for the ability to modify config-data on products without replicating packages.
Note: Setting this preference to True will keep reposado from downloading packages unless you also specify OnlyRewriteDeprecatedURLs.

Credit to @weswhet for AlwaysRewriteDistributionURLs :)

### CODE REVIEW

**Comment by gregneagle** [comment]
I don't think I understand the comment/statement "This preference is intended to allow you to save bandwidth." What bandwidth is saved? reposado still has to download all products -- it can't wait until a product is marked as deprecated to download it; by then it's too late.

Local bandwidth is usually cheaper/faster than bandwidth to Apple's servers and is a major reason people have used software update replicas and Caching Services.

Maybe your scenario is reposado repo in some cloud provider and you want to save bandwidth/data transfer cost payments to the cloud provider? If so, clarifying the scenarios in which this would be useful would be helpful.

**Comment by gregneagle** [comment]
As for the patch itself -- the logic looks hard to follow (lots of complex nested if/thens), so it may take me a while to understand and review.

Assuming I agree to merge this, Walt Disney Animation Studios will need a signed Contributor License Agreement from you. I'd point you to one to start filling out, but it looks like they broke something and I need to alert the proper people to get it fixed.

**Comment by sphen13** [comment]
Yes Greg you are right in your thoughts.  I can clarify my meaning in the docs.  In my current use of this I am reducing the amount of client download bandwidth billed from my cloud provider.  It is not aimed at reducing the bandwidth required on the client end (thats the job of a caching proxy in my case).

I am further using a script that only syncs my chosen "deprecated" items up to my cloud repo - significantly reducing the amount of storage needed on the cloud platform.

Either way I thought this would be a handy thing to share (and write a blog post explaining my use of)

**Comment by sphen13** [comment]
Thanks Greg - I will look out for that.  The logic is a bit hard to follow (was for me when I was writing).  It think this had to do with combining it with the `AlwaysRewriteDistributionURLs` option as well.

I will help as much as I can.  I appreciate your time.

**Comment by gregneagle** [comment]
Sorry it's been a while; I'll probably not have the time to properly review for a little bit longer. In the meantime, could you submit a Contributor License Agreement?
https://github.com/wdas/reposado/blob/master/CONTRIBUTING.md#contributor-license-agreement
I can't accept contributions without one on file for you.

**Comment by sphen13** [comment]
Thanks greg - stupid question but is Schedule B required? should I describe the patch or can it be blank and be a blanket for any contribution?  thx

**Comment by gregneagle** [comment]
I am not a lawyer. I think you could just write "enhancements to reposado".

**Comment by sphen13** [comment]
:) thx - submitted.
